### ISSUE TITLE
Failed to carry Concat or Mul ( version 0.9)

### ISSUE THREAD

**Issue by zltgo** [gorgonia/gorgonia#341]
```
import (
	"fmt"
	"gorgonia.org/tensor"

	"github.com/pkg/errors"
	. "gorgonia.org/gorgonia"
)


func concat(xS, yS tensor.Shape){
	
        g := NewGraph()
        x := NewTensor(g, Float32, xS.Dims(), WithName("x"), WithShape(xS...), WithInit(RangedFrom(0)))
        y := NewTensor(g, Float32, yS.Dims(), WithName("y"), WithShape(yS...), WithInit(RangedFrom(0)))

        z, err := Concat(1, x, y)
        if err != nil {
		fmt.Println(errors.Wrap(err, "Concat"))
		return
	}

        m := NewLispMachine(g, ExecuteFwdOnly())
        if err = m.RunAll(); err != nil {
		fmt.Println(errors.Wrap(err, "run Concat"))
		return
        }

        fmt.Print("x: \n", x.Value())
        fmt.Print("\ny: \n", y.Value())
        fmt.Print("\nConcat of x,y: \n", z.Value())
}

func main() {
	
        //success
        concat(tensor.Shape{2, 2, 2}, tensor.Shape{2, 3, 2})

        //run Concat: RunAll: Failed to bindVar: Failed to carry op.Do(): Failed to perform Concat:
        //Unable to assignArray in denseConcat: BroadcastStrides failed: Dimension mismatch. Expected 2, got 3
        concat(tensor.Shape{2, 2, 2}, tensor.Shape{2, 1, 2})
}

```

**Comment by zltgo** [comment]
- The Mul function also panics in same case.

import (
	"fmt"

	"github.com/pkg/errors"
	. "gorgonia.org/gorgonia"
	"gorgonia.org/tensor"
)

func runMul(xS, yS tensor.Shape) {
	
        g := NewGraph()
        x := NewMatrix(g, Float32, WithName("x"), WithShape(xS...), WithInit(RangedFrom(0)))
        y := NewMatrix(g, Float32, WithName("y"), WithShape(yS...), WithInit(RangedFrom(0)))

        z := Must(Mul(x, y))
        cost := Must(Sum(z))

        _, err := Grad(cost, x, y)
        if err != nil {
		fmt.Println(errors.Wrap(err, "Grad"))
		return
        }

        m := NewTapeMachine(g)
        if err = m.RunAll(); err != nil {
		fmt.Println(errors.Wrap(err, "run Concat"))
		return
        }
        fmt.Println("cost:", cost.Value())
}

func main() {

	//it works, cost = 91
	runMul(tensor.Shape{2, 3}, tensor.Shape{3, 2})

	//panic
	runMul(tensor.Shape{2, 2}, tensor.Shape{2, 1})

	//panic
	runMul(tensor.Shape{1, 2}, tensor.Shape{2, 2})
}

panic: runtime error: index out of range

goroutine 1 [running]:
gorgonia.org/gorgonia.linAlgBinOp.InferShape(0x10000, 0xc00019aa80, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/op_math.go:551 +0xe5e
gorgonia.org/gorgonia.ApplyOp(0x7fbff36861f8, 0xc0001af6e7, 0xc0001a2d80, 0x2, 0x2, 0x0, 0x0, 0x0)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/op.go:200 +0x3f7
gorgonia.org/gorgonia.binOpNode(0xbd84e0, 0xc0001af6e7, 0xc0001b8fc0, 0xc0001b8b60, 0x0, 0x0, 0x0)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/operations.go:43 +0x2c7
gorgonia.org/gorgonia.matMulDiffExpr(0xbc0000, 0xc0001b8a80, 0xc0001b8b60, 0xc0001b8c40, 0xc0001b8fc0, 0xc0001af6e0, 0x8, 0x18, 0xc00020b820, 0xbd69a0)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/operatorLinAlg.go:90 +0x184
gorgonia.org/gorgonia.linAlgBinOp.SymDiff(0x0, 0xc0001a2b00, 0x2, 0x2, 0xc0001b8c40, 0xc0001b8fc0, 0x4098cd, 0x20, 0xa8b9e0, 0xbd6401, ...)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/op_math.go:609 +0x103
gorgonia.org/gorgonia.Backpropagate(0xc0001a8178, 0x1, 0x1, 0xc0001a8180, 0x1, 0x1, 0xc0001a2ba0, 0x2, 0x2, 0x0, ...)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/differentiation.go:296 +0xbf1
gorgonia.org/gorgonia.Grad(0xc0001b8d20, 0xc0001a2ba0, 0x2, 0x2, 0xc0001b8d20, 0x0, 0x0, 0x0, 0x0)
	/home/zyx/mygo/src/gorgonia.org/gorgonia/gorgonia.go:237 +0x2c6
main.runMul(0xc00013df78, 0x2, 0x2, 0xc00013df68, 0x2, 0x2)
	/home/zyx/mygo/src/github.com/zltgo/deepmind/issues/mul.go:19 +0x32b
main.main()
	/home/zyx/mygo/src/github.com/zltgo/deepmind/issues/mul.go:49 +0xc9

**Comment by owulveryck** [comment]
Hi;
I created a branch issue-341 to track this issue.
I've created a test file with both of your test cases for investigation.

**Comment by zltgo** [comment]
@owulveryck  
thanks

**Comment by phillips96** [comment]
Hello, has this been looked at since? I'm noticing the same error when attempting to use the `concat` op, which is kicking up errors when back-propagating in the `reduceAdd` of gradients.

Thanks.

**Comment by chewxy** [comment]
hmmt this shouldn't be the case @phillips96 . I have fixed this some time back but may have rebroken it. Can I have an example code?

**Comment by phillips96** [comment]
I've emailed over, along with the work for the other issue.

**Comment by chewxy** [comment]
OK, so thanks to @phillips96 I have determined that this was fixed, but then #359 broke it again, leading to gorgonia/tensor#57. I'll be working on that tonight. Expect a fix in a couple of days.

**Comment by chewxy** [comment]
To properly fix this, I'll need to update the go.mod file of this to use the latest version of `tensor`. Please hold
